File: src/lowlevel/server/icefsoperators/icefsSymLink.go
/*
 * @Author: Tan90degrees tangentninetydegrees@gmail.com
 * @Date: 2023-03-11 07:18:32
 * @LastEditors: Tan90degrees tangentninetydegrees@gmail.com
 * @LastEditTime: 2023-05-10 09:20:35
 * @FilePath: /icefs/src/lowlevel/server/icefsoperators/icefsSymLink.go
 * @Description:
 *
 * Copyright (C) 2023 Tan90degrees <tangentninetydegrees@gmail.com>.
 */
package icefsoperators

import (
	"context"
	"icefs-server/icefserror"
	pb "icefs-server/icefsgrpc"
	"icefs-server/icefsthrift"

	"golang.org/x/sys/unix"
)

func (s *IcefsServer) doIcefsSymLink(parentFakeInode uint64, link string, name string, fuseEntryParamBuilder FuseEntryParamBuilder) (status int32, entry any) {
	var err error
	s.inodeCacheLock.RLock()
	inode := s.getIcefsInode(parentFakeInode)
	if inode == nil {
		s.inodeCacheLock.RUnlock()
		status = icefserror.ICEFS_BUG_ERR
		return
	}

	inode.inodeLock.RLock()
	s.inodeCacheLock.RUnlock()
	status = icefserror.IcefsStdErrno(unix.Symlinkat(link, inode.fd, name))
	inode.inodeLock.RUnlock()

	entry, err = s.doIcefsLookUp(parentFakeInode, name, fuseEntryParamBuilder)
	status = icefserror.IcefsStdErrno(err)
	return
}

func (s *IcefsGRpcServer) DoIcefsSymLink(ctx context.Context, req *pb.IcefsSymLinkReq) (*pb.IcefsSymLinkRes, error) {
	var res pb.IcefsSymLinkRes
	var entry any

	res.Status, entry = s.server.doIcefsSymLink(req.ParentInode, req.Link, req.Name, GRpcFuseEntryParamBuilder)
	if res.Status == icefserror.ICEFS_EOK {
		res.Entry = entry.(*pb.FuseEntryParam)
	}

	return &res, nil
}

func (s *IcefsThriftServer) DoIcefsSymLink(ctx context.Context, req *icefsthrift.IcefsSymLinkReq) (*icefsthrift.IcefsSymLinkRes, error) {
	var res icefsthrift.IcefsSymLinkRes
	var entry any

	res.Status, entry = s.server.doIcefsSymLink(uint64(req.ParentInode), req.Link, req.Name, ThriftFuseEntryParamBuilder)
	if res.Status == icefserror.ICEFS_EOK {
		res.Entry = entry.(*icefsthrift.FuseEntryParam)
	}

	return &res, nil
}

File: src/lowlevel/client/icefsoperators/icefsSymLink.cpp
/*
 * @Author: Tan90degrees tangentninetydegrees@gmail.com
 * @Date: 2023-03-30 04:19:29
 * @LastEditors: Tan90degrees tangentninetydegrees@gmail.com
 * @LastEditTime: 2023-04-21 02:41:06
 * @FilePath: /icefs/src/lowlevel/client/icefsoperators/icefsSymLink.cpp
 * @Description:
 *
 * Copyright (C) 2023 Tan90degrees <tangentninetydegrees@gmail.com>.
 */
#include <stdio.h>

#include <string>

#include "icefsClient.hpp"
#include "icefsClientUtils.hpp"

void IcefsClient::DoIcefsSymLink(fuse_req_t fuseReq, const char *link,
                                 fuse_ino_t parentInode, const char *name) {
  ICEFS_PR_FUNCTION;

  switch (this->clientConfig.linkType) {
    case ICEFS_LINK_USE_GRPC: {
      icefsgrpc::IcefsSymLinkReq req;
      icefsgrpc::IcefsSymLinkRes res;
      grpc::ClientContext ctx;
      struct fuse_entry_param entry;

      req.set_link(link);
      req.set_parent_inode(parentInode);
      req.set_name(name);

      grpc::Status status = gRpcClient->DoIcefsSymLink(&ctx, req, &res);
      if (status.ok() && !res.status()) {
        IcefsGRpcFillFuseEntryParamIn(&entry, res.entry());
        fuse_reply_entry(fuseReq, &entry);
      } else {
        fuse_reply_err(fuseReq, res.status() ? res.status() : EIO);
        ICEFS_PR_ERR_STATUS;
      }
      break;
    }

    case ICEFS_LINK_USE_THRIFT: {
      icefsthrift::IcefsSymLinkReq req;
      icefsthrift::IcefsSymLinkRes res;
      struct fuse_entry_param entry;

      req.__set_link(link);
      req.__set_parent_inode(parentInode);
      req.__set_name(name);

      icefsThriftConn *thriftConn = thriftClientPool->GetIcefsThriftConn();
      thriftConn->thriftClient->DoIcefsSymLink(res, req);
      thriftClientPool->PutIcefsThriftConn(thriftConn);
      if (!res.status) {
        IcefsThriftFillFuseEntryParamIn(&entry, res.entry);
        fuse_reply_entry(fuseReq, &entry);
      } else {
        fuse_reply_err(fuseReq, res.status);
        ICEFS_PR_ERR_STATUS;
      }
      break;
    }

    default:
      fuse_reply_err(fuseReq, EIO);
      break;
  }
}
